### PR TITLE
[GPT-fast benchmark] Add MLP, gather + gemv, gemv micro benchmark

### PR DESCRIPTION
**Pull request by yanboliang**
Output example:
```
| name                         | metric                    | target  | actual  |
|------------------------------|---------------------------|---------|---------|
| layer_norm_bfloat16          | memory_bandwidth(GB/s)    | 1017    | 1000.01 |
| mlp_layer_norm_gelu_bfloat16 | flops_utilization         | 0.71    | 0.71    |
| gemv_int8                    | memory_bandwidth(GB/s)    | 990     | 984.06 |
| gemv_bfloat16                | memory_bandwidth(GB/s)    | 1137    | 1137.92 |
| gather_gemv_int8             | memory_bandwidth(GB/s)    | 1113    | 1111.09 |
| gather_gemv_bfloat16         | memory_bandwidth(GB/s)    | 1249    | 1248.15 |


```

### CODE REVIEW

File: benchmarks/gpt_fast/benchmark.py [hunk 1]
```diff
@@ -18,57 +21,172 @@ class Experiment:
...
 
-        if i == -1:
-            print(f"Compilation time: {time.perf_counter() - t0:.2f} seconds")
-            continue
+            def gather_gemv(W, score_idxs, x):
+                return W[score_idxs].to(x.dtype) @ x
 
-        torch.cuda.synchronize("cuda")
-        total_time += time.perf_counter() - t0
+            W = torch.randn(E, D, D, device="cuda").to(dtype=dtype)
```
Comment (Chillee) [inline benchmarks/gpt_fast/benchmark.py:129]: Transpose the last 2 dimensions - I think that would be faster.

File: benchmarks/gpt_fast/benchmark.py [hunk 2]
```diff
@@ -18,57 +21,172 @@ class Experiment:
...
 
-    total_time = total_time / num_samples
+            compiled_fn = torch.compile(gather_gemv)
 
-    return total_time
+            for _ in range(WARMUP_ITER):
+                compiled_fn(W, score_idxs, x)
 
+            us_per_iter = do_bench(lambda: compiled_fn(W, score_idxs, x)) * 1000
+            memory_bandwidth += (1e6 / us_per_iter) * 4 * D * D * dtype.itemsize / 1e9
```
Comment (Chillee) [inline benchmarks/gpt_fast/benchmark.py:139]: Why is this `4 * D* D`? I think it should be 2 * D * D.

File: benchmarks/gpt_fast/benchmark.py [hunk 3]
```diff
@@ -18,57 +21,172 @@ class Experiment:
...
+            x = torch.randn(D, device="cuda", dtype=torch.bfloat16)
+
+            compiled_mod = torch.compile(mod)
+
+            for _ in range(WARMUP_ITER):
+                compiled_mod(x)
+
+            us_per_iter = do_bench(lambda: compiled_mod(x)) * 1000
+            memory_bandwidth += (
+                (1e6 / us_per_iter) * 4 * D * intermediate_size * dtype.itemsize / 1e9
```
Comment (Chillee) [inline benchmarks/gpt_fast/benchmark.py:63]: I don't get how this is calculatred. For SimpleMLP I think it's better to report flops achieved or something like that.

**Review by Chillee — CHANGES_REQUESTED** [review]

File: benchmarks/gpt_fast/benchmark.py (hunk 2 above)
Comment (yanboliang) [reply benchmarks/gpt_fast/benchmark.py:139]: both read and write?

File: benchmarks/gpt_fast/benchmark.py (hunk 3 above)
Comment (yanboliang) [reply benchmarks/gpt_fast/benchmark.py:63]: Do you know which is the standard helper function that we used to get flops like ```triton.do_bench```?

Comment (yanboliang) [reply benchmarks/gpt_fast/benchmark.py:63]: Oh, nvm, I think it just need to calculate it by my hand.

File: benchmarks/gpt_fast/benchmark.py (hunk 2 above)
Comment (yanboliang) [reply benchmarks/gpt_fast/benchmark.py:139]: Addressed offline.

File: benchmarks/gpt_fast/benchmark.py (hunk 3 above)
Comment (Chillee) [reply benchmarks/gpt_fast/benchmark.py:63]: You can use `torch.utils.flop_counter`

**Review by Chillee — APPROVED** [review]

**Pull request closed without merging**